### PR TITLE
op-node: Fix reset loop

### PR DESCRIPTION
**Pull request by trianglesphere**
**Description**

This fixes a reset loop caused by state being incorrectly cleared. The first commit
does the simple fix & adds a regression test. The second does a more complete fix.

**Metadata**

- Fixes CLI-3763

**TODOs**

- [x] Author or reviewer has added an entry to the [current release notes draft][RND], if appropriate.

[RND]: https://www.notion.so/oplabs/ded30107ceec41c88817e60322aa8d0a?v=b4a22cedb85a46a38c9be14e7c984953&pvs=4 "Release Notes"

### CODE REVIEW

**Review by protolambda — APPROVED** [review]
Change LGTM, but need to clean up the TODO, I think we should keep the condition mentioned there.

File: op-node/rollup/derive/engine_queue.go [hunk 1]
```diff
@@ -430,15 +436,17 @@ func (eq *EngineQueue) tryNextSafeAttributes(ctx context.Context) error {
 		return nil
 	}
 	// validate the safe attributes before processing them. The engine may have completed processing them through other means.
-	if eq.safeHead != eq.safeAttributesParent {
-		if eq.safeHead.ParentHash != eq.safeAttributesParent.Hash {
+	if eq.safeHead != eq.safeAttributes.parent {
+		// TODO: when is this condition true? It appears to be so in tests, but shouldn't be true otherwise.
```
[the remark this answers is not in the document]
Comment (trianglesphere) [reply op-node/rollup/derive/engine_queue.go:440]: The TODO is the other way. If the first condition is true `eq.safeHead != eq.safeAttributesParent` that should imply that `eq.safeHead.Hash != eq.safeAttributesParent.Hash`. I don't think we ever hit the log `queued safe attributes are stale`

**Comment by trianglesphere** [comment]
> And to confirm my understanding - when we reset the engine queue, we may wind up discarding a set of attributes we'd previously pulled but not yet applied to the execution engine (the `safeAttributes` value). That's ok though because the whole pipeline will be reset back to be in sync with the execution engine and thus we'll wind up pulling those attributes from L1 again and so we'll come back round to applying them again if they're still relevant. If the reset was because of an L1 reorg, we may wind up not getting those attributes from the new L1 chain or they may now be discarded because they don't match our new L1 origin.

Yep that is correct. This has an additional change of removing the queue attributes on a reset.
The original bug was related to not clearing `eq.safeAttributesParent` on a reset.

File: op-node/rollup/derive/engine_queue.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (trianglesphere) [reply op-node/rollup/derive/engine_queue.go]: Ok, I think I finally figured it out. I inverted the condition here @protolambda and left a comment on what I think is happening.

File: op-node/rollup/derive/engine_queue.go [hunk 2]
```diff
@@ -430,15 +436,19 @@ func (eq *EngineQueue) tryNextSafeAttributes(ctx context.Context) error {
...
 	}
 	// validate the safe attributes before processing them. The engine may have completed processing them through other means.
-	if eq.safeHead != eq.safeAttributesParent {
-		if eq.safeHead.ParentHash != eq.safeAttributesParent.Hash {
-			return NewResetError(fmt.Errorf("safe head changed to %s with parent %s, conflicting with queued safe attributes on top of %s",
-				eq.safeHead, eq.safeHead.ParentID(), eq.safeAttributesParent))
+	if eq.safeHead != eq.safeAttributes.parent {
+		// Previously the attribute's parent was the safe head. If the safe head advances so safe head's parent is the same as the
+		// attribute's parent then we need to cancel the attributes.
+		if eq.safeHead.ParentHash == eq.safeAttributes.parent.Hash {
```
Comment (Inphi) [inline op-node/rollup/derive/engine_queue.go:442]: Is this check still needed now that we're always considering the `safeAttributes` during resets? How does the discrepancy described here still occur? I couldn't find a test that reaches L443.

Comment (trianglesphere) [reply op-node/rollup/derive/engine_queue.go:442]: L443 is reached in the block building race test.

**Pull request merged**